topdown/strings.go

```diff
@@ -234,27 +233,31 @@ func builtinReplace(a, b, c ast.Value) (ast.Value, error) {
 }
 
 func builtinReplaceN(a, b ast.Value) (ast.Value, error) {
-	asJSON, err := ast.JSON(a)
+	oldnewObj, err := builtins.ObjectOperand(a, 1)
 	if err != nil {
 		return nil, err
 	}
-	oldnewObj, ok := asJSON.(map[string]interface{})
-	if !ok {
-		return nil, builtins.NewOperandTypeErr(1, a, "object")
-	}
 
 	s, err := builtins.StringOperand(b, 2)
 	if err != nil {
 		return nil, err
 	}
 
 	var oldnewArr []string
-	for k, v := range oldnewObj {
-		strVal, ok := v.(string)
+	err = oldnewObj.Iter(func(k *ast.Term, v *ast.Term) error {
```

Comment:
The object implementation iterates in insertion order so there could still be different outputs with this change. What about getting the keys from the object, sorting them, and then iterating over the sorted key set?

Reply:
🤔 As in `strings.replace_n(o1, s)` could differ from `strings.replace_n(o2, s)` although o1 and o2 are the same in terms of key-val pairs? Yup, and that is unfortunate. I had only considered multiple evals of `strings.replace_n(o1, s)` which should now no longer have different outcomes.

Sorting them would guarantee stable output, of course, but it would also mean that there's no way for the user to control the replacement order. I think this comes down to the interface of `strings.replace_n` being unfortunate, it's using something unordered, so you cannot control the ordering, end of story.

Since nobody has ever been complaining about this, it's probably not a big deal, the common "foo -> bar, baz -> quz" usecase isn't affected.

tl;dr: I'll update this, sorting the keys, tomorrow. Controlling the replacement ordering is just not going to be supported. (Or stay unsupported, rather.)

Reply:
Technically, we could go back to using `ast.JSON`, and then sort the keys; but I think I like `builtins.ObjectOperand` better. WDYT?

Reply:
> tl;dr: I'll update this, sorting the keys, tomorrow. Controlling the replacement ordering is just not going to be supported. (Or stay unsupported, rather.)

Sounds good. We can add that in the future if someone requests it.

> Technically, we could go back to using ast.JSON, and then sort the keys; but I think I like builtins.ObjectOperand better. WDYT?

I'd probably just use the `ast` types. I'd call `Keys()` and then sort the result.

Reply:
✅  Updated